### PR TITLE
Refactor revm_spec function to simplify hardfork detection logic

### PR DESCRIPTION
**Pull request by cypherpepe**
Replaced individual hardfork checks in revm_spec with a loop-based approach using find. Moved hardfork-to-SpecId mapping logic to a separate helper function (map_hardfork_to_spec_id) for cleaner and more maintainable code.

### CODE REVIEW

File: crates/node/src/evm.rs [hunk 1]
```diff
@@ -252,51 +252,43 @@ impl ConfigureEvm for OdysseyEvmConfig {
...
-        reth_revm::primitives::HOMESTEAD
-    } else if chain_spec.fork(EthereumHardfork::Frontier).active_at_head(block) {
-        reth_revm::primitives::FRONTIER
-    } else {
-        panic!(
-            "invalid hardfork chainspec: expected at least one hardfork, got {:?}",
-            chain_spec.hardforks
+    chain_spec
+        .hardforks
+        .iter()
```
Comment (onbjerg) [inline crates/node/src/evm.rs:257]: please use `forks_iter()`

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/node/src/evm.rs [hunk 2]
```diff
@@ -252,51 +252,43 @@ impl ConfigureEvm for OdysseyEvmConfig {
...
+                    chain_spec.hardforks
+                )
+            },
+            |(fork, _)| map_hardfork_to_spec_id(fork)
         )
+}
+
+
+/// Map hardfork to the corresponding `SpecId`.
+fn map_hardfork_to_spec_id(fork: &EthereumHardfork) -> reth_revm::primitives::SpecId {
```
Comment (onbjerg) [inline crates/node/src/evm.rs:271]: some of these are `OptimismHardfork`, which means we'd need dynamic dispatch. not really sure this is the way to go

Comment (cypherpepe) [reply crates/node/src/evm.rs:271]: Given the issue with handling both EthereumHardfork and OptimismHardfork, would you recommend simplifying revm_spec by removing map_hardfork_to_spec_id altogether and directly matching on each fork type? Here’s what that would look like:

```rust
fn revm_spec(chain_spec: &ChainSpec, block: &Head) -> SpecId {
    chain_spec
        .forks_iter()
        .find(|(_, condition)| condition.active_at_head(block))
        .map_or_else(
            || {
                panic!(
                    "Invalid chainspec: expected at least one active hardfork, got {:?}",
                    chain_spec.hardforks
                );
            },
            |(fork, _)| {
                if let Some(eth_fork) = fork.downcast_ref::<EthereumHardfork>() {
                    match eth_fork {
                        EthereumHardfork::Prague => SpecId::OSAKA,
                        // (other EthereumHardfork mappings)
                    }
                } else if let Some(opt_fork) = fork.downcast_ref::<OptimismHardfork>() {
                    match opt_fork {
                        OptimismHardfork::Granite => SpecId::GRANITE,
                        // (other OptimismHardfork mappings)
                    }
                } else {
                    panic!("Unsupported hardfork: {:?}", fork);
                }
            }
        )
}
```

This approach would avoid dynamic dispatch and handle both fork types without additional casting. Do you think this is a better solution?

**Review by mattsse — COMMENTED** [review]
I find the if else chain easier to read/maintain

**Comment by cypherpepe** [comment]
Thank you for the feedback. I'll keep your suggestions in mind and continue looking for ways to support the team. Let's close this thread, as these changes are not the right fit.

**Comment by mattsse** [comment]
appreciate it

**Pull request closed without merging**